python/lsst/daf/butler/registries/oracleRegistry.py

```diff
@@ -0,0 +1,116 @@
+
+__all__ = ("OracleRegistry", )
+
+
+# This is special cased as oracle does not support IF NOT EXISTS,
+# but should be kept until the consequences of using CREATE OR REPLACE
+# would be on users who are actively using a view when someone else
+# attempts to create or replace it.
+def _onOracleConnect(dbapiConnection, connectionRecord):
+    ifExistFunction = """
+    CREATE OR REPLACE PROCEDURE
+    view_exists_guard (name IN VARCHAR2, stmt IN LONG)
+    IS
+        nCount NUMBER;
+    BEGIN
+        SELECT count(view_name) into nCount FROM user_views where view_name = name;
+        IF(ncount <= 0)
+        THEN
+            execute immediate stmt;
+        END IF;
+    END view_exists_guard;
+    """
+    with closing(dbapiConnection.cursor()) as cursor:
+        cursor.execute(ifExistFunction)
+
+
+class OracleRegistry(SqlRegistry):
+    """Registry backed by a Oracle database.
+
+    Parameters
+    ----------
+    config : `SqlRegistryConfig` or `str`
+        Load configuration
+    """
+
+    @classmethod
+    def setConfigRoot(cls, root, config, full):
+        """Set any filesystem-dependent config options for this Registry to
+        be appropriate for a new empty repository with the given root.
+
+        Parameters
+        ----------
+        root : `str`
+            Filesystem path to the root of the data repository.
+        config : `Config`
+            A Butler-level config object to update (but not a
+            `ButlerConfig`, to avoid included expanded defaults).
+        full : `ButlerConfig`
+            A complete Butler config with all defaults expanded;
+            repository-specific options that should not be obtained
+            from defaults when Butler instances are constructed
+            should be copied from `full` to `Config`.
+        """
+        super().setConfigRoot(root, config, full)
+        Config.overrideParameters(RegistryConfig, config, full,
+                                  toCopy=("cls", "deferDatasetIdQueries"))
+
+    def __init__(self, registryConfig, schemaConfig, dimensionConfig, create=False):
+        registryConfig = SqlRegistryConfig(registryConfig)
+        self.schemaConfig = schemaConfig
+        super().__init__(registryConfig, schemaConfig, dimensionConfig, create)
+
+    def _createEngine(self):
+        tables = self.schemaConfig['tables'].keys()
+
+        def _ignoreQuote(name, matchObj):
+            text = matchObj.group(0)
+            if text.startswith("'") and text.endswith("'"):
+                return text
+            elif text.startswith('"') and text.endswith('"'):
+                return text
+            else:
+                return text.replace(name, f'"{name}"')
+
+        def _oracleExecute(connection, cursor, statement, parameters, context, executemany):
+            for name in tables:
+                if name in statement:
+                    statement = re.sub(f"\"?'?{name}\\b'?\"?", lambda x: _ignoreQuote(name, x), statement)
```

Comment:
I'm not very sure what the purpose of this code is? Does it try to fix some hand-written queries with incorrectly quoted or unquoted table names? Can you add a comment? This really looks like something that we should not be doing (hack is probably a better name for it) and it does not feel very robust. If the problem is with our hand-written queries then we should fix those.

Reply:
I will add a comment to the code, but basically the oracle database as it is currently setup does not support our case sensitive names of our tables correctly. This means all table names need to be quoted (in certain contexts). This rewrites all statements to quote the tables appropriately. Hopefully long term this will not be necessary as they are working to remove this issue. I will say as much in the comment.

Reply:
If it's not only for hand-written queries but for general case then I'm not happy that we are fighting SQLAlchemy rules. Would be easier to just use SQLAlchemy consistently everywhere than trying to workaround it. I'm OK for it for now but it should be revisited and discussed at some point (I'm not sure I understand all issues here).